### PR TITLE
Moving ConcurrentQueue to shared folder

### PR DESCRIPTION
**Pull request by maryamariyan**
corefx PR: dotnet/corefx#29643

Fixes coreclr issue: #17751 

Related to this [comment](https://github.com/dotnet/coreclr/pull/17800#issuecomment-385092192) made by @stephentoub:
> That just leaves ConcurrentQueue, which is used by ThreadPool. I agree that one is reasonable to move down.

cc: @stephentoub @jkotas @safern @danmosemsft

### CODE REVIEW

File: src/System.Private.CoreLib/shared/System/Collections/Concurrent/ConcurrentQueue_Segment.cs [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+        {
+            /// <summary>The item.</summary>
+            public T Item;
+            /// <summary>The sequence number for this slot, used to synchronize between enqueuers and dequeuers.</summary>
+            public int SequenceNumber;
+        }
+    }
+
+    /// <summary>A placeholder class for common padding constants and eventually routines.</summary>
+    internal static class PaddingHelpers
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Collections/Concurrent/ConcurrentQueue_Segment.cs]: This seems to be duplicated with src\System.Private.CoreLib\src\Internal\Padding.cs

File: src/System.Private.CoreLib/src/System/Collections/Concurrent/IProducerConsumerCollection.cs [hunk 2]
```diff
@@ -2,29 +2,21 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/System/Collections/Concurrent/IProducerConsumerCollection.cs]: This file should be move to shared as well.

File: src/System.Private.CoreLib/src/System/Collections/Concurrent/SystemCollectionsConcurrent_ProducerConsumerCollectionDebugView.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+using System.Diagnostics;
+
+namespace System.Collections.Concurrent
+{
+    /// <summary>
+    /// A debugger view of the IProducerConsumerCollection that makes it simple to browse the
+    /// collection's contents at a point in time.
+    /// </summary>
+    /// <typeparam name="T">The type of elements stored within.</typeparam>
+    internal sealed class SystemCollectionsConcurrent_ProducerConsumerCollectionDebugView<T>
```
Comment (jkotas) [inline src/System.Private.CoreLib/src/System/Collections/Concurrent/SystemCollectionsConcurrent_ProducerConsumerCollectionDebugView.cs]: This should be in shared as well, and named just `ProducerConsumerCollectionDebugView`

**Comment by maryamariyan** [comment]
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

File: src/System.Private.CoreLib/System.Private.CoreLib.csproj [hunk 4]
```diff
@@ -626,6 +623,10 @@
   <ItemGroup>
     <Compile Include="src\System\PinnableBufferCacheEventSource.cs" />
     <Compile Include="src\System\Runtime\RuntimeImports.cs" />
+    <Compile Include="shared\System\Collections\Concurrent\ConcurrentQueue.cs" />
+    <Compile Include="shared\System\Collections\Concurrent\ConcurrentQueue_Segment.cs" />
+    <Compile Include="shared\System\Collections\Concurrent\IProducerConsumerCollection.cs" />
+    <Compile Include="shared\System\Collections\Concurrent\ProducerConsumerCollectionDebugView.cs" />
```
Comment (maryamariyan) [inline src/System.Private.CoreLib/System.Private.CoreLib.csproj]: I had to add these files to the csproj to fix compile issues with the project. Please let me know if this is OK.

Comment (jkotas) [reply src/System.Private.CoreLib/System.Private.CoreLib.csproj]: I think you are still going to have compile issues.

The shared files should not be included from this project file.

I think the problem is that you have these files in Windows specific section in the shared project file. Try moving them to the platform neutral section.

**Comment by maryamariyan** [comment]
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

**Pull request merged**